File: User/ugcp_svc/Content/X52/CWinUSBX52.cpp
#include "../framework.h"
#include <SetupAPI.h>
#include <hidsdi.h>
#include "CWinUSBX52.h"
#include "MFDMenu.h"


bool CWinUSBX52::Prepare()
{
    Close(false);

    HDEVINFO diDevs = INVALID_HANDLE_VALUE;
    SP_DEVICE_INTERFACE_DATA diData{};

    diDevs = SetupDiGetClassDevs(&guidInterface, NULL, NULL, (DIGCF_PRESENT | DIGCF_DEVICEINTERFACE));
    if (INVALID_HANDLE_VALUE == diDevs)
    {
        return false;
    }

    diData.cbSize = sizeof(SP_DEVICE_INTERFACE_DATA);
    if (!SetupDiEnumDeviceInterfaces(diDevs, NULL, &guidInterface, 0, &diData))
    {
        DWORD err = GetLastError();
        SetupDiDestroyDeviceInfoList(diDevs);
        return (err == ERROR_NO_MORE_ITEMS);
    }

    DWORD size = 0;
    if ((FALSE == SetupDiGetDeviceInterfaceDetail(diDevs, &diData, NULL, 0, &size, NULL)) && (ERROR_INSUFFICIENT_BUFFER != GetLastError()))
    {
        SetupDiDestroyDeviceInfoList(diDevs);
        return false;
    }

    auto buf = std::make_unique<std::uint8_t[]>(size);
    memset(buf.get(), 0, size);
    PSP_DEVICE_INTERFACE_DETAIL_DATA didData = reinterpret_cast<PSP_DEVICE_INTERFACE_DETAIL_DATA>(buf.get());
    didData->cbSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA);
    if (!SetupDiGetDeviceInterfaceDetail(diDevs, &diData, didData, size, &size, NULL))
    {
        SetupDiDestroyDeviceInfoList(diDevs);
        return false;
    }

    {
        std::lock_guard<std::mutex> lock(mutex);
        if (hardwareId == HARDWARE_ID_X52)
        {
            pathInterface = didData->DevicePath; //string is copied
        }
    }

    SetupDiDestroyDeviceInfoList(diDevs);

    return true;
}

bool CWinUSBX52::Open()
{
    std::unique_lock<std::mutex> lock(mutex);
    {
        if ((hardwareId != 0) && (!pathInterface.empty()) && (hwusb.load() == nullptr))
        {
            unique_handle nhdev(CreateFileW(pathInterface.c_str(), GENERIC_WRITE | GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED, NULL));
            if (!nhdev.valid())
            {
                return false;
            }
            else
            {
                WINUSB_INTERFACE_HANDLE wih = nullptr;
                if (!WinUsb_Initialize(nhdev.get(), &wih))
                {
                    lock.unlock();
                    Close(false);
                    return false;
                }

                hdev.set(nhdev.move()); //hdev should be nullptr here
                hwusb.store(wih);

                memset(&pipe, 0, sizeof(WINUSB_PIPE_INFORMATION));
                if (!WinUsb_QueryPipe(wih, 0, 0, &pipe))
                {
                    lock.unlock();
                    Close(false);
                    return false;
                }

                DWORD newReportLenght = 0;
                if (HidD_GetPreparsedData(nhdev.get(), reinterpret_cast<PHIDP_PREPARSED_DATA*>(&preparsed)))
                {
                    HIDP_CAPS caps;
                    if (HidP_GetCaps(reinterpret_cast<PHIDP_PREPARSED_DATA>(preparsed), &caps) == HIDP_STATUS_SUCCESS)
                    {
                        if (GetDeviceMap(preparsed, &caps))
                        {
                            newReportLenght = caps.InputReportByteLength;
                        }
                    }
                }
                if (newReportLenght != 0)
                {
                    reportLenght.store(newReportLenght);
                    reportBuffer = std::make_unique<CHAR[]>(newReportLenght);
                    CMFDMenu::Get().SetWelcome();
                }
                else
                {
                    lock.unlock();
                    Close(false);
                    return false;
                }
            }
        }
    }

    return true;
}

void CWinUSBX52::Close(bool exit)
{
    {
        std::unique_lock<std::mutex> lock(mutex);
        HANDLE h = hwusb.exchange(nullptr);
        if (h != nullptr)
        {
            CancelIoEx(h, NULL);
            WinUsb_Free(h);
        }
    }

    CHIDDevices::Close(exit);
}

unsigned short CWinUSBX52::Read(void* buff)
{
    if (!paused.load())
    {
        auto hUSB = hwusb.load();
        if (hUSB == nullptr)
        {
            std::this_thread::sleep_for(std::chrono::milliseconds(3000));
            bool prepared;
            bool exit;
            {
                std::unique_lock<std::mutex> lock(mutex);
                prepared = !pathInterface.empty();
                exit = (hardwareId == 0);
            }
            if (exit)
            {
                return 0;
            }
            if (!prepared)
            {
                Prepare();
            }
            Open();
            return 0;
        }
        else
        {
            ULONG readSize = 0;
            static_cast<char*>(buff)[0] = 0; //report id
            if (WinUsb_ReadPipe(hUSB, pipe.PipeId, &static_cast<std::uint8_t*>(buff)[1], reportLenght.load() - 1, &readSize, nullptr))
            {
                readSize++;
                ULONG size = 0;
                if (HidP_GetData(HidP_Input, nullptr, &size, reinterpret_cast<PHIDP_PREPARSED_DATA>(preparsed), reportBuffer.get(), reportLenght.load()) == HIDP_STATUS_BUFFER_TOO_SMALL)
                {
                    if (HidP_GetData(HidP_Input, reinterpret_cast<PHIDP_DATA>(buff), &size, reinterpret_cast<PHIDP_PREPARSED_DATA>(preparsed), reportBuffer.get(), reportLenght.load()) == HIDP_STATUS_SUCCESS)
                    {
                        return static_cast<unsigned short>(size);
                    }
                }
                return 0;
            }
        }
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(1500));
    return 0;
}

void CWinUSBX52::SetPause(bool onoff)
{
    if (onoff)
    {
        if (paused.exchange(true) == false)
        {
            Close(false);
        }
    }
    else
    {
        Close(false);
        paused.store(false);
    }
}

File: User/ugcp_svc/X52/CWinUSBX52.cs
using System;
using System.Runtime.InteropServices;
using API;

namespace ugcp_svc.X52
{
    sealed class CWinUSBX52 : HIDInput.CHIDDevice
    {
        //constexpr auto HARDWARE_ID_X52 = L"\\\\?\\USB#VID_06A3&PID_0255";
        public const uint HARDWARE_ID_X52 = 0x063a0255;
        private Guid guidInterface = new("{ 0xA57C1168, 0x7717, 0x4AF0, { 0xB3, 0x0E, 0x6A, 0x4C, 0x62, 0x30, 0xBB, 0x10 }}");

        private nint _hwusb = nint.Zero;
        private nint HwUSB
        {
            get
            {
                return System.Threading.Interlocked.CompareExchange(ref _hwusb, nint.Zero, nint.Zero);
            }
            set
            {
                System.Threading.Interlocked.Exchange(ref _hwusb, value);
                CX52Write.Get()?.SetWinUSB(value);
            }
        }
        CWinUSB.WINUSB_PIPE_INFORMATION pipe = new();
        private byte paused;

        public CWinUSBX52(Profile.CProfile refProfile) : base(HARDWARE_ID_X52, refProfile)
        {
        }

        protected override bool Prepare()
        {
            Close();

            nint diDevs;
            SetupDi.SP_DEVICE_INTERFACE_DATA diData = new();

            diDevs = SetupDi.SetupDiGetClassDevsW(ref guidInterface, null, nint.Zero, SetupDi.DIGCF_PRESENT | SetupDi.DIGCF_DEVICEINTERFACE);
            if (diDevs == Win32.INVALID_HANDLE_VALUE)
            {
                return false;
            }

            diData.cbSize = Marshal.SizeOf<SetupDi.SP_DEVICE_INTERFACE_DATA>();
            if (!SetupDi.SetupDiEnumDeviceInterfaces(diDevs, nint.Zero, ref guidInterface, 0, ref diData))
            {
                SetupDi.SetupDiDestroyDeviceInfoList(diDevs);
                return false;
            }

            if (!SetupDi.SetupDiGetDeviceInterfaceDetailW(diDevs, ref diData, nint.Zero, 0, out uint size, nint.Zero) && (Marshal.GetLastWin32Error() != 122)) //122 = ERROR_INSUFFICIENT_BUFFER
            {
                SetupDi.SetupDiDestroyDeviceInfoList(diDevs);
                return false;
            }

            using Shared.RAII.HGlobalHandle buf = new((int)size);
            Marshal.WriteInt32(buf.DangerousGetHandle(), 8); //x64
            if (!SetupDi.SetupDiGetDeviceInterfaceDetailW(diDevs, ref diData, buf.DangerousGetHandle(), size, out size, nint.Zero))
            {
                SetupDi.SetupDiDestroyDeviceInfoList(diDevs);
                return false;
            }

            string? ninterface = Marshal.PtrToStringAuto(buf.DangerousGetHandle() + 4);
            if (ninterface == null)
            {
                SetupDi.SetupDiDestroyDeviceInfoList(diDevs);
                return false;
            }

            lock(mutex)
            {
                if (HardwareId == HARDWARE_ID_X52)
                {
                    pathInterface = ninterface;
                }
            }

            SetupDi.SetupDiDestroyDeviceInfoList(diDevs);
            return true;
        }

        protected override bool Open()
        {
            lock (mutex)
            {
                if ((HardwareId != 0) && !string.IsNullOrEmpty(pathInterface) && (HwUSB == nint.Zero))
                {
                    using Shared.RAII.UniqueHandle nhdev = new(Win32.CreateFileW(pathInterface, Win32.GENERIC_READ | Win32.GENERIC_WRITE, Win32.FILE_SHARE_READ | Win32.FILE_SHARE_WRITE, nint.Zero, Win32.OPEN_EXISTING, 0x80 | 0x40000000, nint.Zero)); // 0x80 | 0x40000000 = FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED
                    if (nhdev.IsInvalid)
                    {
                        return false;
                    }

                    if (!CWinUSB.WinUsb_Initialize(nhdev.DangerousGetHandle(), out nint wih))
                    {
                        if (wih != nint.Zero)
                        {
                            CWinUSB.WinUsb_Free(wih);
                        }
                        return false;
                    }

                    pipe = new();
                    if (!CWinUSB.WinUsb_QueryPipe(wih, 0, 0, ref pipe))
                    {
                        CWinUSB.WinUsb_Free(wih);
                        return false;
                    }

                    uint newReportLenght = 0;
                    nint newPrepased = nint.Zero;
                    if (HID.HidD_GetPreparsedData(nhdev.DangerousGetHandle(), ref newPrepased))
                    {
                        using Shared.RAII.HGlobalHandle caps = new(Marshal.SizeOf<HID.HIDP_CAPS>());
                        if (HID.HidP_GetCaps(newPrepased, caps.DangerousGetHandle()) == HID.HIDP_STATUS_SUCCESS)
                        {
                            if (GetDeviceMap(newPrepased, caps.DangerousGetHandle()))
                            {
                                HID.HIDP_CAPS mcaps = Marshal.PtrToStructure<HID.HIDP_CAPS>(caps.DangerousGetHandle());
                                newReportLenght = mcaps.InputReportByteLength;
                            }
                        }
                    }
                    if (newReportLenght != 0)
                    {
                        reportBuffer = new byte[newReportLenght];
                        preparsed = newPrepased;
                        hdev = new(nhdev.Move());
                        HwUSB = wih;
                        CMFDMenu.Get()?.SetWelcome();
                    }
                    else
                    {
                        CWinUSB.WinUsb_Free(wih);
                        return false;
                    }
                }
            }

            return true;
        }

        protected override void Close()
        {
            lock(mutex)
            {
                nint old = HwUSB;
                HwUSB = nint.Zero;
                if (old != nint.Zero)
                {
                    Win32.CancelIoEx(old, nint.Zero);
                    CWinUSB.WinUsb_Free(old);
                }
            }
            base.Close();
        }

        public override ushort Read(Span<byte> buff)
        {
            if (System.Threading.Interlocked.CompareExchange(ref paused, 0, 0) == 1)
            {
                System.Threading.Tasks.Task.Delay(1500).Wait();
                return 0;
            }

            bool opened;
            bool prepared;
            lock (mutex)
            {
                if (HardwareId == 0) //disposing
                {
                    return 0;
                }
                prepared = !string.IsNullOrEmpty(pathInterface);
                opened = !hdev.IsInvalid;
            }
            if (!prepared)
            {
                Prepare();
            }
            if (!opened)
            {
                Open();
                System.Threading.Tasks.Task.Delay(1500).Wait();
                return 0;
            }

            lock (mutex)
            {
                if (!hdev.IsInvalid)
                {
                    uint readSize = 0;
                    buff[0] = 0;
                    if (CWinUSB.WinUsb_ReadPipe(HwUSB, pipe.PipeId, buff.Slice(1), (uint)(reportBuffer.Length - 1), ref readSize, nint.Zero))
                    {
                        readSize++;
                        uint size = 0;
                        if (HID.HidP_GetData(0, nint.Zero, ref size, preparsed, reportBuffer, (uint)reportBuffer.Length) == HID.HIDP_STATUS_BUFFER_TOO_SMALL)
                        {
                            if (HID.HidP_GetData(0, buff, ref size, preparsed, reportBuffer, (uint)reportBuffer.Length) == HID.HIDP_STATUS_SUCCESS)
                            {
                                return (ushort)size;
                            }
                        }
                    }
                }
            }

            System.Threading.Tasks.Task.Delay(500).Wait();
            return 0;
        }

        public void SetPause(bool onoff)
        {
            if (onoff)
            {
                System.Threading.Interlocked.Exchange(ref paused, 1);
            }
            else
            {
                System.Threading.Interlocked.Exchange(ref paused, 0);
            }
        }
    }
}
